File: app/src/main/java/com/kickstarter/viewmodels/FacebookConfirmationViewModel.java
package com.kickstarter.viewmodels;

import android.util.Pair;

import com.kickstarter.libs.ActivityViewModel;
import com.kickstarter.libs.CurrentConfigType;
import com.kickstarter.libs.CurrentUserType;
import com.kickstarter.libs.Environment;
import com.kickstarter.services.ApiClientType;
import com.kickstarter.services.apiresponses.AccessTokenEnvelope;
import com.kickstarter.services.apiresponses.ErrorEnvelope;
import com.kickstarter.ui.IntentKey;
import com.kickstarter.ui.activities.FacebookConfirmationActivity;

import androidx.annotation.NonNull;
import rx.Notification;
import rx.Observable;
import rx.subjects.BehaviorSubject;
import rx.subjects.PublishSubject;

import static com.kickstarter.libs.rx.transformers.Transformers.combineLatestPair;
import static com.kickstarter.libs.rx.transformers.Transformers.errors;
import static com.kickstarter.libs.rx.transformers.Transformers.takeWhen;
import static com.kickstarter.libs.rx.transformers.Transformers.values;

public interface FacebookConfirmationViewModel {

  interface Inputs {
    /** Call when the create new account button has been clicked. */
    void createNewAccountClick();

    /** Call when the send newsletter switch has been toggled. */
    void sendNewslettersClick(boolean __);
  }

  interface Outputs {
    /** Fill the view's email address. */
    Observable<String> prefillEmail();

    /** Emits a string to display when sign up fails. */
    Observable<String> signupError();

    /** Finish Facebook confirmation activity with OK result. */
    Observable<Void> signupSuccess();

    /** Emits a boolean to check send newsletter switch. */
    Observable<Boolean> sendNewslettersIsChecked();
  }

  final class ViewModel extends ActivityViewModel<FacebookConfirmationActivity> implements Inputs, Outputs {
    private final ApiClientType client;
    private final CurrentUserType currentUser;
    private final CurrentConfigType currentConfig;

    public ViewModel(final @NonNull Environment environment) {
      super(environment);

      this.client = environment.apiClient();
      this.currentConfig = environment.currentConfig();
      this.currentUser = environment.currentUser();

      final Observable<String> facebookAccessToken = intent()
        .map(i -> i.getStringExtra(IntentKey.FACEBOOK_TOKEN))
        .ofType(String.class);

      final Observable<Pair<String, Boolean>> tokenAndNewsletter = facebookAccessToken
        .compose(combineLatestPair(this.sendNewslettersIsChecked));

      intent()
        .map(i -> i.getParcelableExtra(IntentKey.FACEBOOK_USER))
        .ofType(ErrorEnvelope.FacebookUser.class)
        .map(ErrorEnvelope.FacebookUser::email)
        .compose(bindToLifecycle())
        .subscribe(this.prefillEmail::onNext);

      final Observable<Notification<AccessTokenEnvelope>> createNewAccountNotification = tokenAndNewsletter
        .compose(takeWhen(this.createNewAccountClick))
        .flatMap(tn -> this.client.registerWithFacebook(tn.first, tn.second))
        .share()
        .materialize();

      createNewAccountNotification
        .compose(errors())
        .map(ErrorEnvelope::fromThrowable)
        .map(ErrorEnvelope::errorMessage)
        .takeUntil(this.signupSuccess)
        .compose(bindToLifecycle())
        .subscribe(this.signupError);

      createNewAccountNotification
        .compose(values())
        .ofType(AccessTokenEnvelope.class)
        .compose(bindToLifecycle())
        .subscribe(this::registerWithFacebookSuccess);

      this.sendNewslettersClick
        .compose(bindToLifecycle())
        .subscribe(this.sendNewslettersIsChecked::onNext);

      this.currentConfig.observable()
        .take(1)
        .map(config -> false)
        .subscribe(this.sendNewslettersIsChecked::onNext);
    }

    private void registerWithFacebookSuccess(final @NonNull AccessTokenEnvelope envelope) {
      this.currentUser.login(envelope.user(), envelope.accessToken());
      this.signupSuccess.onNext(null);
    }

    private final PublishSubject<Void> createNewAccountClick = PublishSubject.create();
    private final PublishSubject<Boolean> sendNewslettersClick = PublishSubject.create();

    private final BehaviorSubject<String> prefillEmail = BehaviorSubject.create();
    private final PublishSubject<String> signupError = PublishSubject.create();
    private final PublishSubject<Void> signupSuccess = PublishSubject.create();
    private final BehaviorSubject<Boolean> sendNewslettersIsChecked = BehaviorSubject.create();

    public final Inputs inputs = this;
    public final Outputs outputs = this;

    @Override public void createNewAccountClick() {
      this.createNewAccountClick.onNext(null);
    }
    @Override public void sendNewslettersClick(final boolean b) {
      this.sendNewslettersClick.onNext(b);
    }

    @Override public @NonNull Observable<String> prefillEmail() {
      return this.prefillEmail;
    }
    @Override public @NonNull Observable<String> signupError() {
      return this.signupError;
    }
    @Override public @NonNull Observable<Void> signupSuccess() {
      return this.signupSuccess;
    }
    @Override public @NonNull Observable<Boolean> sendNewslettersIsChecked() {
      return this.sendNewslettersIsChecked;
    }
  }
}

File: app/src/main/java/com/kickstarter/viewmodels/FacebookConfirmationViewModel.kt
package com.kickstarter.viewmodels

import android.content.Intent
import android.os.Parcelable
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.kickstarter.libs.Environment
import com.kickstarter.libs.rx.transformers.Transformers
import com.kickstarter.libs.utils.extensions.addToDisposable
import com.kickstarter.libs.utils.extensions.isNotNull
import com.kickstarter.services.apiresponses.AccessTokenEnvelope
import com.kickstarter.services.apiresponses.ErrorEnvelope
import com.kickstarter.services.apiresponses.ErrorEnvelope.FacebookUser
import com.kickstarter.ui.IntentKey
import com.kickstarter.viewmodels.usecases.LoginUseCase
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.BehaviorSubject
import io.reactivex.subjects.PublishSubject

interface FacebookConfirmationViewModel {
    interface Inputs {
        /** Call when the create new account button has been clicked.  */
        fun createNewAccountClick()

        /** Call when the send newsletter switch has been toggled.  */
        fun sendNewslettersClick(b: Boolean)
    }

    interface Outputs {
        /** Fill the view's email address.  */
        fun prefillEmail(): Observable<String>

        /** Emits a string to display when sign up fails.  */
        fun signupError(): Observable<String>

        /** Finish Facebook confirmation activity with OK result.  */
        fun signupSuccess(): Observable<Unit>

        /** Emits a boolean to check send newsletter switch.  */
        fun sendNewslettersIsChecked(): Observable<Boolean>
    }

    class FacebookConfirmationViewModel(environment: Environment) :
        ViewModel(), Inputs, Outputs {
        private val client = requireNotNull(environment.apiClientV2())
        private val loginUserCase = LoginUseCase(environment)
        private val currentConfig = requireNotNull(environment.currentConfigV2())

        private fun registerWithFacebookSuccess(envelope: AccessTokenEnvelope) {
            loginUserCase.setToken(envelope.accessToken())
            loginUserCase.setUser(envelope.user())
            signupSuccess.onNext(Unit)
        }

        private val createNewAccountClick: PublishSubject<Unit> = PublishSubject.create()
        private val sendNewslettersClick: PublishSubject<Boolean> = PublishSubject.create()

        private val prefillEmail: BehaviorSubject<String> = BehaviorSubject.create()
        private val signupError: PublishSubject<String> = PublishSubject.create()
        private val signupSuccess: PublishSubject<Unit> = PublishSubject.create()
        private val sendNewslettersIsChecked: BehaviorSubject<Boolean> = BehaviorSubject.create()
        private val intent = PublishSubject.create<Intent>()
        private val disposables = CompositeDisposable()

        @JvmField
        val inputs: Inputs = this
        @JvmField
        val outputs: Outputs = this

        init {
            val intentObservable = intent.share()

            val facebookAccessToken = intentObservable
                .map { i: Intent -> i.getStringExtra(IntentKey.FACEBOOK_TOKEN) }
                .ofType(String::class.java)

            val tokenAndNewsletter = facebookAccessToken
                .compose(Transformers.combineLatestPair(this.sendNewslettersIsChecked))

            intentObservable
                .filter { it.getParcelableExtra<Parcelable>(IntentKey.FACEBOOK_USER).isNotNull() }
                .map { it.getParcelableExtra<Parcelable>(IntentKey.FACEBOOK_USER) }
                .filter { it is FacebookUser }
                .map { it as FacebookUser }
                .map { it.email() }
                .subscribe { prefillEmail.onNext(it) }
                .addToDisposable(disposables)

            val createNewAccountNotification = tokenAndNewsletter
                .compose(Transformers.takeWhenV2(this.createNewAccountClick))
                .flatMap {
                    client.registerWithFacebook(
                        it.first, it.second
                    )
                }
                .materialize()

            createNewAccountNotification
                .compose(Transformers.errorsV2())
                .map { ErrorEnvelope.fromThrowable(it) }
                .map { it.errorMessage() }
                .takeUntil(this.signupSuccess)
                .subscribe { this.signupError.onNext(it) }
                .addToDisposable(disposables)

            createNewAccountNotification
                .compose(Transformers.valuesV2())
                .ofType(AccessTokenEnvelope::class.java)
                .subscribe { envelope: AccessTokenEnvelope ->
                    this.registerWithFacebookSuccess(
                        envelope
                    )
                }
                .addToDisposable(disposables)

            sendNewslettersClick
                .subscribe { v: Boolean -> sendNewslettersIsChecked.onNext(v) }
                .addToDisposable(disposables)

            currentConfig.observable()
                .take(1)
                .map { false }
                .subscribe { v: Boolean -> sendNewslettersIsChecked.onNext(v) }
                .addToDisposable(disposables)
        }

        override fun createNewAccountClick() {
            createNewAccountClick.onNext(Unit)
        }

        override fun sendNewslettersClick(b: Boolean) {
            sendNewslettersClick.onNext(b)
        }

        override fun prefillEmail(): Observable<String> {
            return this.prefillEmail
        }

        override fun signupError(): Observable<String> {
            return this.signupError
        }

        override fun signupSuccess(): Observable<Unit> {
            return this.signupSuccess
        }

        override fun sendNewslettersIsChecked(): Observable<Boolean> {
            return this.sendNewslettersIsChecked
        }

        fun provideIntent(intent: Intent) {
            this.intent.onNext(intent)
        }

        override fun onCleared() {
            disposables.clear()
            super.onCleared()
        }
    }

    class Factory(private val environment: Environment) : ViewModelProvider.Factory {
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            return FacebookConfirmationViewModel(environment) as T
        }
    }
}
